Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,560 @@
+pattern ::= pattern '?'
+pattern ::= pattern '{' number (',' spaces? number?)? '}'
+```
+
+| Syntax      | Description           | Capture type                                                  |
+| ----------- | --------------------- | ------------------------------------------------------------- |
+| `*`         | 0 or more             | `Array` of the sub-pattern capture type                       |
+| `+`         | 1 or more             | `Array` of the sub-pattern capture type                       |
+| `?`         | 0 or 1                | `Optional` of the sub-pattern capture type                    |
+| `{n}`       | Exactly _n_           | `Array` of the sub-pattern capture type                       |
+| `{n, m}`    | Between _n_ and _m_   | `Array` of the sub-pattern capture type                       |
+| `{n,}`      | _n_ or more           | `Array` of the sub-pattern capture type                       |
+
+```swift
+/([0-9a-fA-F]+)+/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     OneOrMore {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)*/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)?/
+// => `Regex<Substring?>`
+
+// Result builder DSL equivalent:
+//     Optionally {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == Substring?`
+
+/([0-9a-fA-F]+){3}/
+// => `Regex<[Substring]>
+
+// Result builder DSL equivalent:
+//     Repeat(3) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,5}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...5) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+```
+
+Note:
+- For quantifiers that produce an array, it is arguable that a lazy collection
+  based on matched ranges could minimize reference counting operations on
+  `Substring` and reduce allocations. However, array in capture types would make
+  a much cleaner type signature.
+- For repetitions or an exact count, e.g. `[a-z]{2}`, it would slightly improve
+  type safety to make its capture type be a homogeneous tuple instead of an
+  array. However, because large homogeneous tuples are difficult to work with
+  and because fixed-count repetitions are syntactically very similar to
+  dynamic-count repetitions which produce an array, an tuple capture may come as
+  a surprise. An array on the other is a better choise for both ease of
+  use and consistency.
```

Comment:
Do we want a note that lazy quantifiers don't affect the capture type? 

Maybe something like:
```
- Lazy quantifiers have the same capture type semantics as their greedy counterparts.
  If the first example literal from above were instead written as `([0-9a-fA-F]+)+?`,
  it would yield the same `Regex<[Substring]>` type.
```

Reply:
Good point.

Reply:
Should these notes be moved to subsections of Alternatives Considered?